frontend/app/src/next/components/runs/runs-table/data-table.tsx

```diff
@@ -0,0 +1,254 @@
+'use client';
+
+import { useState, useMemo, useCallback, useEffect } from 'react';
+import {
+  ColumnDef,
+  ColumnFiltersState,
+  VisibilityState,
+  RowSelectionState,
+  OnChangeFn,
+  flexRender,
+  getCoreRowModel,
+  getFacetedRowModel,
+  getFacetedUniqueValues,
+  getFilteredRowModel,
+  getSortedRowModel,
+  useReactTable,
+  Row,
+  ExpandedState,
+} from '@tanstack/react-table';
+import {
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+  TableRow,
+} from '@/next/components/ui/table';
+import { cn } from '@/next/lib/utils';
+import { useNavigate } from 'react-router-dom';
+import { ROUTES } from '@/next/lib/routes';
+import { V1WorkflowType } from '@/lib/api';
+
+const styles = {
+  status: 'p-0 w-[40px]',
+  runId: 'border-r border-border',
+};
+
+interface IDGetter {
+  metadata: {
+    id: string;
+  };
+  isExpandable?: boolean;
+}
+
+interface DataTableProps<TData extends IDGetter, TValue> {
+  columns: ColumnDef<TData, TValue>[];
+  data: TData[];
+  emptyState?: React.ReactNode;
+  isLoading?: boolean;
+  selectedTaskId?: string;
+  onRowClick?: (row: TData) => void;
+  onSelectionChange?: (selectedRows: TData[]) => void;
+  rowSelection?: RowSelectionState;
+  setRowSelection?: OnChangeFn<RowSelectionState>;
+  selectAll?: boolean;
+  getSubRows?: (originalRow: TData, index: number) => TData[];
+}
+
+export function DataTable<TData extends IDGetter, TValue>({
+  columns,
+  data,
+  emptyState,
+  isLoading,
+  selectedTaskId,
+  onRowClick = () => {},
+  onSelectionChange,
+  rowSelection = {},
+  setRowSelection,
+  selectAll = false,
+  getSubRows,
+}: DataTableProps<TData, TValue>) {
+  const navigate = useNavigate();
+  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>([]);
+  const [columnVisibility, setColumnVisibility] = useState<VisibilityState>({});
+  const [expanded, setExpanded] = useState<ExpandedState>({});
+
+  const memoizedRowSelection = useMemo(() => {
+    if (selectAll) {
+      return data.reduce((acc, _, index) => ({ ...acc, [index]: true }), {});
+    }
+    return rowSelection;
+  }, [selectAll, data, rowSelection]);
+
+  const table = useReactTable({
+    data,
+    columns,
+    state: {
+      columnFilters,
+      columnVisibility,
+      rowSelection: memoizedRowSelection,
+      expanded,
+    },
+    onColumnFiltersChange: setColumnFilters,
+    onColumnVisibilityChange: setColumnVisibility,
+    onRowSelectionChange: setRowSelection,
+    enableRowSelection: true,
+    getCoreRowModel: getCoreRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    getFilteredRowModel: getFilteredRowModel(),
+    getFacetedRowModel: getFacetedRowModel(),
+    getFacetedUniqueValues: getFacetedUniqueValues(),
+    getRowId: (row) => {
+      const typedRow = row as { taskExternalId?: string; id?: string };
+      return typedRow.taskExternalId || typedRow.id || String(Math.random());
```

Comment:
I think this is potentially a source of bugs in the future - I feel like we should guarantee that `row` always has an `id` or `externalId` or something and definitely not fall back to a random string, but could be convinced otherwise

Reply:
we'll let this one bite us later ;)